akka-remote/src/main/scala/akka/remote/artery/tcp/TcpFraming.scala

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+package tcp
+
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+
+import akka.annotation.InternalApi
+import akka.remote.artery.FlightRecorderEvents.TcpInbound_Received
+import akka.stream.Attributes
+import akka.stream.impl.io.ByteStringParser
+import akka.stream.impl.io.ByteStringParser.ByteReader
+import akka.stream.impl.io.ByteStringParser.ParseResult
+import akka.stream.impl.io.ByteStringParser.ParseStep
+import akka.stream.scaladsl.Framing.FramingException
+import akka.stream.stage.GraphStageLogic
+import akka.util.ByteString
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object TcpFraming {
+  val Undefined = Int.MinValue
+
+  /**
+   * The first 4 bytes of a new connection must be these `64x0 75x0 75x0 64x0` (AKKA).
+   * The purpose of the "magic" is to detect and reject weird (accidental) accesses.
+   */
+  val Magic = ByteString('A'.toByte, 'K'.toByte, 'K'.toByte, 'A'.toByte)
+
+  /**
+   * When establishing the connection this header is sent first.
+   * It contains a "magic" and the stream identifier for selecting control, ordinary, large
+   * inbound streams.
+   *
+   * The purpose of the "magic" is to detect and reject weird (accidental) accesses.
+   * The magic 4 bytes are `64x0 75x0 75x0 64x0` (AKKA).
+   *
+   * The streamId` is encoded as 1 byte.
+   */
+  def encodeConnectionHeader(streamId: Int): ByteString =
+    Magic ++ ByteString(streamId.toByte)
+
+  /**
+   * Each frame starts with the frame header that contains the length
+   * of the frame. The `frameLength` is encoded as 4 bytes (little endian).
+   */
+  def encodeFrameHeader(frameLength: Int): ByteString =
+    ByteString(
+      (frameLength & 0xff).toByte,
+      ((frameLength & 0xff00) >> 8).toByte,
+      ((frameLength & 0xff0000) >> 16).toByte,
+      ((frameLength & 0xff000000) >> 24).toByte
+    )
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class TcpFraming(flightRecorder: EventSink) extends ByteStringParser[EnvelopeBuffer] {
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new ParsingLogic {
+    abstract class Step extends ParseStep[EnvelopeBuffer]
+    startWith(ReadMagic)
+
+    case object ReadMagic extends Step {
+      override def parse(reader: ByteReader): ParseResult[EnvelopeBuffer] = {
+        val magic = ByteString(reader.readByte(), reader.readByte(), reader.readByte(), reader.readByte())
+        if (magic == TcpFraming.Magic)
+          ParseResult(None, ReadStreamId)
+        else
+          throw new FramingException("Stream didn't start with expected magic bytes. " +
```

Comment:
Probably not worth logging what was actually sent

Reply:
I think it might be actually nice to output the first 10 bytes or so. E.g. as hex as `(magic ++ reader.remainingData).take(10).map(_ formatted "%02x").mkString(" ")`.

Reply:
ok, I was thinking that we shouldn't be accused of logging unknown user data (you always send your password as first bytes in a connection, right?), but ok that's too paranoid.